### DISCUSSION TITLE
Default values are not automatically mapped when using decorators.

### DISCUSSION THREAD

**Discussion by rowanparker** [mikro-orm/mikro-orm#2582] [Q&A]
https://mikro-orm.io/docs/defining-entities#default-values

Under the second method, it says that the default value will be `automatically mapped to the entity property after it is being persisted (after flush)`.

This does not seem to work for me. I am using the following entities.

```
@Entity()
export class Book extends BaseEntity<any, any> {
  @PrimaryKey()
  id!: number;

  @Property()
  title!: string;

  @OneToMany('BookGenre', 'book')
  genres = new Collection<BookGenre>(this);
}

@Entity()
export class Genre extends BaseEntity<any, any> {
  @PrimaryKey()
  id!: number;

  @Property()
  name!: string;

  @OneToMany('BookGenre', 'genre')
  books = new Collection<BookGenre>(this);
}

@Entity()
export class BookGenre extends BaseEntity<any, any> {
  @PrimaryKey()
  id!: number;

  @ManyToOne()
  book!: Book;

  @ManyToOne()
  genre!: Genre;

  @Property({ default: 50 })
  votes;
}
```

I create 3 objects, then call persist and flush on the `book`. This **is** saved to the database correctly.

```
const book = new Book();
book.title = 'Some book';

const genre = new Genre();
genre.name = 'Some genre';

const bookGenre = new BookGenre();
bookGenre.book = book;
bookGenre.genre = genre;

await orm.em.persistAndFlush(book);

return Promise.resolve(book);
```

However when I debug and inspect the book object, immediately after the flush, the `votes` value is missing from the BookGenre object (but is in the database table).

![image](https://user-images.githubusercontent.com/331336/147458646-c8de05e4-324d-4138-bdb2-f144f4b62e61.png)

Is this a bug or am I reading the documentation wrong?

I think it is similar to this issue:
https://github.com/mikro-orm/mikro-orm/issues/2391

**Comment by B4nan** [comment]
You should see the reloading query in logs, or an insert query with returning statement (in postgres), what are the logs saying?

I can confirm it works with postgres, but not with mysql/sqlite, which is quite weird, I remember having tests for that somewhere.

> I think it is similar to this issue: #2391

That issue is about very different thing (runtime values) on quite diffferent level (embeddables). 

Btw next time If you want to report bugs, please don't ignore the issue template - what versions are you on? What driver? 

Anyway, workaround for now (which is what I would use in the first place, as runtime defaults should be preferred for additional consistency even before you flush):

```ts
  @Property({ default: 50 })
  votes: number = 50;
```

**Comment by rowanparker** [comment]
> what versions are you on? What driver?

Apologies for not using the correct template. Versions in use:

```
"@mikro-orm/cli": "^4.5.10",
"@mikro-orm/core": "^4.5.10",
"@mikro-orm/migrations": "^4.5.10",
"@mikro-orm/nestjs": "^4.3.1",
"@mikro-orm/sqlite": "^4.5.10",
```

This output above is from a new project I created to reproduce this issue. I first noticed it in a large project with these versions:

```
"@mikro-orm/cli": "^4.5.7",
"@mikro-orm/core": "^4.5.7",
"@mikro-orm/migrations": "~4.5.9",
"@mikro-orm/mysql": "^4.5.9",
"@mikro-orm/nestjs": "^4.2.0",
"@mikro-orm/reflection": "^4.5.9",
"@mikro-orm/sql-highlighter": "^1.0.1",
```

> You should see the reloading query in logs, or an insert query with returning statement (in postgres), what are the logs saying?

No subsequent SELECT statement is executed. This is all that appears in the log:

```
begin
insert into `genre` (`name`) values ('Some genre')
insert into `book` (`title`) values ('Some book')
insert into `book_genre` (`book_id`, `genre_id`) values (1, 1)
commit
```

> Anyway, workaround for now (which is what I would use in the first place, as runtime defaults should be preferred for additional consistency even before you flush)

Understood, that will probably be what we do. I would like to help resolve the bug or update the docs - depending on what your opinion is.

**Reply by B4nan** [reply]
It is a bug for sure, its quite weird that nobody reported this earlier, I remember it was working. Will add tests specifically for this feature and check all the drivers, as right now it was tested only as part of other tests. 

We already have a mechanism for reloading version values, so this should be handled same way.

**Reply by B4nan** [reply]
Fixed in latest dev versions (also backported to 4.x branch, that one is currently building so will be available in few minutes).

**Comment by olivierlacan** [comment]
We're having the same issue with an enum column that has a SQL default in Postgres. The enum is for a `status` attribute that is optional so I wrote a test that submits the attributes without the `status`. I wanted to test that the missing argument did in fact result in us getting an object with the SQL column default back, but that's not the case. 

In my test I only get what I expect after doing `em.clear()` and refetching the object from the database with `em.findOneOrFail()`.

This felt odd to me because the semantics of "flush" seemed to imply in my mind that the database persistence would also lead to the object being reloaded with any persisted default values coming from the database (and not the inputs supplied to the ORM). So frankly I suspect you didn't get earlier bug reports because people assumed they had to clear the Entity Manager manually.  

Here's the basic test structure: 

```typescript
// createExperience()
const experience = new Experience(attributes);
wrap(experience).assign(
  { curators: curators.map((curator) => curator.handle) },
  { em: this.entityManager },
);

await this.entityManager.persistAndFlush(experience);

return experience;
```

```typescript
experience = await createExperience({ title: "Example" });

expect(experience).to.have.property("id").to.be.a("string");
expect(experience)
  .to.have.property("title")
  .to.eql("Example");
// this expectation fails
expect(experience)
  .to.have.property("status")
  .to.eql("IN_PROGRESS");

em.clear();
experience = await em.findOneOrFail(Experience, experience.id);

// this expectation succeeds
expect(experience)
  .to.have.property("status")
  .to.eql("IN_PROGRESS");
```

```typescript
// Experience entity
export enum ContentStatus {
  IN_PROGRESS = "IN_PROGRESS",
  PUBLISHED = "PUBLISHED",
  IN_REVISION = "IN_REVISION",
  RETIRED = "RETIRED",
  REMOVED = "REMOVED",
  CANCELLED = "CANCELLED"
}

export class Experience extends BaseEntity {
  // ...
  @Enum({
    columnType: "content_status_type",
    nullable: false,
  })
  status!: ContentStatus;
  // ...
```

```sql
-- Table Definition ----------------------------------------------

CREATE TABLE experiences (
    id uuid DEFAULT uuid_generate_v4() PRIMARY KEY,
    title text NOT NULL,
    status content_status_type NOT NULL DEFAULT 'IN_PROGRESS'::content_status_type
    created_at timestamp with time zone NOT NULL DEFAULT CURRENT_TIMESTAMP,
    updated_at timestamp with time zone NOT NULL DEFAULT CURRENT_TIMESTAMP,
);

-- Indices -------------------------------------------------------

CREATE UNIQUE INDEX experiences_pkey ON experiences(id uuid_ops);
```

We use: 

```
"@mikro-orm/cli": "~4.5.9",
"@mikro-orm/core": "~4.5.9",
"@mikro-orm/migrations": "~4.5.9",
"@mikro-orm/nestjs": "~4.3.1",
"@mikro-orm/postgresql": "~4.5.9",
```

The issue is pretty clear with debug logs: 

```
[query] insert into "experiences" ("created_at", "title", "updated_at") values ('2022-01-07T08:55:36.618Z', 'Example', '2022-01-07T08:55:36.618Z') returning "id", "created_at", "updated_at"
```


**Update**:

And this is about when I realized our `status` @Enum was missing a `defaultRaw` or `default` attribute declaration which is why it wasn't included in the `RETURNING` statement, unlike `id` and our timestamp columns. Lesson learned. It does feel odd to have to re-specify database-side defaults in order to trigger a refetch on `INSERT` but I don't really have a better solution honestly.  😅

**Reply by B4nan** [reply]
> It does feel odd to have to re-specify database-side defaults in order to trigger a refetch on INSERT but I don't really have a better solution honestly. 😅

How should the ORM know there is a database default, if you don't specify it in the metadata?

**Reply by olivierlacan** [reply]
> How should the ORM know there is a database default, if you don't specify it in the metadata?

Oh, not arguing about that. I'm just stating that the way the `default` specification triggers *RETURNING* is unintuitive. Not that I think there's another way to specify defaults.

**Reply by tukusejssirs** [reply]
I agree with @olivierlacan that _the way the `default` specification triggers RETURNING is unintuitive_, as not all columns need to have defaults.

I think one should be able to define the return types in one of the following ways:
1. [default] use `default` property (this is the way we can do it right now);
2. define to return all columns (e.g. `returnAllColumns: boolean` in the table metadata);
3. define explicitly which columns to return.

For my point of view, having at least the 2nd option added to MikroORM would make it much easier to some users (including me), however, the 3rd option could be useful to others too.

**Reply by ejose19** [reply]
I found this thread while looking for a more elegant solution for the same problem. Currently I can use `@Property({ default: null })` to have the db be the only source of value for defaults and have them be returned on flush, however that may hint the property is nullable which in this case it's not, I'd expect something like `@Property({ databaseDefault: true })` to make it more clear.

I also agree with @tukusejssirs, a toggle on table metadata and persist/flush methods would be ideal to handle some special cases.

**Reply by B4nan** [reply]
So you use `default: null` just to disable the property from `returning` statement? Why do you even care about this? It needs to be there so we can load the value, otherwise your entity would be inconsistent.

It feels like you guys are too concerned with an implementation detail. You need to provide enough metadata for the ORM to work. It is an implementation detail that we use `returning` statement instead of a separate query (like other drivers that don't support `returning` statements do). Having another configuration (that would be postgres specific) seems like it would make things even more unintuitive. If you want defaults from database, you need to provide them in the entity definition. 

I don't understand how would be helpful if we use `returning *` or similar. You won't be able to work with the returned values anyhow - once again, its an implementation detail, if we do `returning *` we would still use only the same properties we know are needed for rehydration. And for me we already have the third option, having separate boolean toggle would only make things more verbose.

Note that I really don't plan to add any options to how the UoW works (so persist/flush methods).

**Reply by ejose19** [reply]
@B4nan No, I use `default: null` to have the database generated value be returned after I persist, currently if I don't specify `default: null` then that property is `undefined`.

here's a small repro on `postgres`:

```sql
CREATE TABLE test (
  a int primary key generated always as identity,
  b text not null default gen_random_uuid()
);
```

```ts
@Entity({ tableName: "test" })
export class Test {
  @PrimaryKey()
  a!: number;

  // Have to use nullable in order for mikro-orm to not throw with
  // `ValidationError  Value for Test.b is required, 'undefined' found`
  @Property({ nullable: true }) 
  b!: string;
}
```

```ts
const x = new Test();
await em.persistAndFlush(x);

console.log(x); // Test { a: 1, b: undefined }
```

notice how `b: undefined`, if I instead used `@Property({ default: null })` then I get the correct result: `Test { a: 2, b: '6dce8d93-d298-4034-8c8c-8f220d3298a7' }`

however, as I said before `default: null` may give the impression that this column is nullable which is not the case, hence something like `@Property({ databaseDefault: true })` would be a better fit

**Reply by B4nan** [reply]
You should put the real default value there, not null.

**Reply by ejose19** [reply]
The problem with doing something like: `defaultRaw: 'gen_random_uuid'` is that now I need to keep both database default and schema default in sync, which is completely unnecessary as (1) I want to keep implementation detail only in database (2) that `defaultRaw` is not even being used (if I remove the default value from the database table, then I get the error `NotNullConstraintViolationException  insert into "test" default values returning "a", "b" - null value in column "b" of relation "test" violates not-null constraint`)

**Reply by ejose19** [reply]
Another reason for a `returning *` is triggers, here's a repro:

```sql
CREATE TABLE test (
  a int primary key generated always as identity,
  b text NOT NULL
);

CREATE OR REPLACE FUNCTION t_test_f()
RETURNS TRIGGER
LANGUAGE plpgsql
AS $$
BEGIN
  NEW.b := 'changed';
  RETURN NEW;
END $$;

CREATE OR REPLACE TRIGGER t_test
BEFORE INSERT OR UPDATE ON test
FOR EACH ROW 
EXECUTE PROCEDURE t_test_f();
```
```ts
@Entity({ tableName: "test" })
export class Test {
  @PrimaryKey()
  a!: number;

  @Property()
  b!: string;

  constructor(b: string) {
    this.b = b;
  }
}
```
```ts
const x = new Test("abc");
await em.persistAndFlush(x);
console.log(x);
```

That returns `Test { a: 1, b: 'abc' }`, however database value is `Test { a: 1, b: 'changed' }`, without a `returning *` or specifying which column to "reload", we would be using invalid values after this point.
